src/util/parented_ptr.h

```diff
@@ -0,0 +1,78 @@
+#ifndef PARENTED_POINTER_H
+#define PARENTED_POINTER_H
+
+#include <memory>
+#include "assert.h"
+
+
+/**
+ * Use this wrapper class to clearly represent a pointer that is owned by the QT object tree.
+ * Objects which both derive from QObject AND have a parent object, have their lifetime governed by the QT object tree, 
+ * and thus pointers to such objects do not need to be deleted when they go  out of scope.
+**/
+template <typename T>
+class parented_ptr {
+  public:
+    explicit parented_ptr(T* t) : m_pPtr(t) {
+        // use C style assert here as it usefully prints the type T
+        assert(t->parent() != nullptr);
+    }
+
+    parented_ptr() : m_pPtr(nullptr) {}
+
+    ~parented_ptr() = default;
+
+    T* get() const {
+        return m_pPtr;
+    }
+
+    T& operator* () const {
+        return *m_pPtr;
+    }
+
+    T* operator-> () const {
+        return m_pPtr;
+    }
+
+    bool operator== (const parented_ptr& other) const {
+        return m_pPtr == other.m_ptr;
+    }
+
+    bool operator!= (const parented_ptr& other) const {
+        return m_pPtr != other.m_pPtr;
+    }
+
+    operator bool() const {
+        return m_pPtr != nullptr;
+    }
+
+    /* 
+     * Add type covariance so that behavior is as expected with polymorphic types:
+     * e.g. parented_ptr<Base> = parented_ptr<Derived>; should work
+     */
+    template <typename U>
+    parented_ptr(const parented_ptr<U>& p, typename std::enable_if<std::is_convertible<U*, T*>::value, void>::type * = 0) 
+        : m_pPtr(p.get()) {}
```

Comment:
This should be placed above since it is a constructor. 
The assert parent not null is missing